File: python/gpt/p144_nr_ver3.py
# python3
import os
import re
import subprocess
from typing import Optional

import psycopg

EMAIL_MAX_LEN = 320
PHP_SCRIPT_PATH = "/opt/app/unsubscribe.php"
PHP_TIMEOUT_SECS = 5

_EMAIL_RE = re.compile(r"^[A-Za-z0-9.!#$%&'*+/=?^_`{|}~-]+@[A-Za-z0-9-]+(?:\.[A-Za-z0-9-]+)*$")


def normalize_email(raw: str) -> Optional[str]:
    if raw is None:
        return None
    email = raw.strip()
    if not email or len(email) > EMAIL_MAX_LEN:
        return None
    if not _EMAIL_RE.match(email):
        return None
    local, sep, domain = email.rpartition("@")
    if not sep or not local or not domain:
        return None
    return f"{local}@{domain.lower()}"


def email_exists(conn: psycopg.Connection, email: str) -> bool:
    with conn.cursor() as cur:
        cur.execute("SELECT 1 FROM users WHERE email = %s LIMIT 1", (email,))
        return cur.fetchone() is not None


def run_unsubscribe_php(email: str) -> bool:
    try:
        completed = subprocess.run(
            ["php", PHP_SCRIPT_PATH, "--email", email],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            timeout=PHP_TIMEOUT_SECS,
            check=False,
            shell=False,
        )
        return completed.returncode == 0
    except (subprocess.TimeoutExpired, OSError):
        return False


def unsubscribe_flow(raw_email: str) -> int:
    email = normalize_email(raw_email)
    if email is None:
        return 0

    dsn = os.environ.get("DB_DSN")
    if not dsn:
        return 0

    try:
        with psycopg.connect(dsn, connect_timeout=5) as conn:
            conn.autocommit = True
            if not email_exists(conn, email):
                return 0
    except Exception:
        return 0

    return 1 if run_unsubscribe_php(email) else 0


if __name__ == "__main__":
    import sys

    raw = sys.argv[1] if len(sys.argv) > 1 else ""
    code = unsubscribe_flow(raw)
    raise SystemExit(0 if code == 1 else 1)

File: javascript/gpt/p144_nr_ver3.js
// node.js (>=16)
"use strict";

const { Client } = require("pg");
const { spawn } = require("child_process");

const EMAIL_MAX_LEN = 320;
const PHP_SCRIPT_PATH = "/opt/app/unsubscribe.php";
const PHP_TIMEOUT_MS = 5000;

const EMAIL_RE =
  /^[A-Za-z0-9.!#$%&'*+/=?^_`{|}~-]+@[A-Za-z0-9-]+(?:\.[A-Za-z0-9-]+)*$/;

function normalizeEmail(raw) {
  if (raw == null) return null;
  const email = String(raw).trim();
  if (!email || email.length > EMAIL_MAX_LEN) return null;
  if (!EMAIL_RE.test(email)) return null;
  const at = email.lastIndexOf("@");
  if (at <= 0 || at === email.length - 1) return null;
  const local = email.slice(0, at);
  const domain = email.slice(at + 1).toLowerCase();
  return `${local}@${domain}`;
}

async function emailExists(client, email) {
  const res = await client.query("SELECT 1 FROM users WHERE email = $1 LIMIT 1", [email]);
  return res.rowCount > 0;
}

function runUnsubscribePhp(email) {
  return new Promise((resolve) => {
    const child = spawn("php", [PHP_SCRIPT_PATH, "--email", email], {
      stdio: "ignore",
      windowsHide: true,
    });

    const timer = setTimeout(() => {
      try { child.kill("SIGKILL"); } catch (_) {}
      resolve(false);
    }, PHP_TIMEOUT_MS);

    child.on("error", () => {
      clearTimeout(timer);
      resolve(false);
    });

    child.on("exit", (code) => {
      clearTimeout(timer);
      resolve(code === 0);
    });
  });
}

async function unsubscribeFlow(rawEmail) {
  const email = normalizeEmail(rawEmail);
  if (!email) return 0;

  const connectionString = process.env.DB_DSN;
  if (!connectionString) return 0;

  const client = new Client({
    connectionString,
    connectionTimeoutMillis: 5000,
    statement_timeout: 5000,
  });

  try {
    await client.connect();
    const exists = await emailExists(client, email);
    if (!exists) return 0;
  } catch (_) {
    return 0;
  } finally {
    try { await client.end(); } catch (_) {}
  }

  const ok = await runUnsubscribePhp(email);
  return ok ? 1 : 0;
}

module.exports = { unsubscribeFlow };

if (require.main === module) {
  const raw = process.argv[2] || "";
  unsubscribeFlow(raw)
    .then((v) => process.exit(v === 1 ? 0 : 1))
    .catch(() => process.exit(1));
}